
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Assert - Node.js - W3cubDocs</title>
  
  <meta name="description" content=" Source Code&#58; lib&#47;assert.js ">
  <meta name="keywords" content="assert, node, js">
  <meta name="HandheldFriendly" content="True">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/node/assert.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-9c7b0d969c0ce639a6c01c2badcdd81667c7c66206aa3d0fccc5cb0585fb1b912dab72ea882101d18010680927514b4b77b107282db8f756fbf63301fe9c88b0.css">
  <script type="text/javascript" src="/assets/application-79c555f6b25481fffac2cac30a7f3e54e608ca09e9e8e42bb1790095ba6d0fcace47d6bc624ddce952c70370892f2d46864f89e6943d4f7f7ff16c8a3231a91a.js"></script>
  <script src="/json/node.js"></script>
  
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-R3WC07G3GB"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-R3WC07G3GB');
</script>
<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2572770204602497"
     crossorigin="anonymous"></script>
<script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/node/" class="_nav-link" title="" style="margin-left:0;">Node.js</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _node">
				
				
<h2 id="assert">Assert
</h2> 
<div class="api_stability api_stability_2">
<a href="https://nodejs.org/api/documentation.html#stability-index">Stability: 2</a> - Stable</div>
 <p><strong>Source Code:</strong> <a href="https://github.com/nodejs/node/blob/v19.0.1/lib/assert.js">lib/assert.js</a></p> <p>The <code>node:assert</code> module provides a set of assertion functions for verifying invariants.</p> <section><h3 id="strict-assertion-mode">Strict assertion mode
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v15.0.0</td> <td><p>Exposed as <code>require('node:assert/strict')</code>.</p></td>
</tr> <tr>
<td>v13.9.0, v12.16.2</td> <td><p>Changed "strict mode" to "strict assertion mode" and "legacy mode" to "legacy assertion mode" to avoid confusion with the more usual meaning of "strict mode".</p></td>
</tr> <tr>
<td>v9.9.0</td> <td><p>Added error diffs to the strict assertion mode.</p></td>
</tr> <tr>
<td>v9.9.0</td> <td><p>Added strict assertion mode to the assert module.</p></td>
</tr> <tr>
<td>v9.9.0</td> <td><p><span>Added in: v9.9.0</span></p></td>
</tr> </tbody>
</table> </details> </div> <p>In strict assertion mode, non-strict methods behave like their corresponding strict methods. For example, <a href="#assertdeepequalactual-expected-message"><code>assert.deepEqual()</code></a> will behave like <a href="#assertdeepstrictequalactual-expected-message"><code>assert.deepStrictEqual()</code></a>.</p> <p>In strict assertion mode, error messages for objects display a diff. In legacy assertion mode, error messages for objects display the objects, often truncated.</p> <p>To use strict assertion mode:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import { strict as assert } from 'node:assert';</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert').strict;</pre> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');</pre> <p>Example error diff:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import { strict as assert } from 'node:assert';

assert.deepEqual([[[1, 2, 3]], 4, 5], [[[1, 2, '3']], 4, 5]);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected ... Lines skipped
//
//   [
//     [
// ...
//       2,
// +     3
// -     '3'
//     ],
// ...
//     5
//   ]</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.deepEqual([[[1, 2, 3]], 4, 5], [[[1, 2, '3']], 4, 5]);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected ... Lines skipped
//
//   [
//     [
// ...
//       2,
// +     3
// -     '3'
//     ],
// ...
//     5
//   ]</pre> <p>To deactivate the colors, use the <code>NO_COLOR</code> or <code>NODE_DISABLE_COLORS</code> environment variables. This will also deactivate the colors in the REPL. For more on color support in terminal environments, read the tty <a href="tty#writestreamgetcolordepthenv"><code>getColorDepth()</code></a> documentation.</p> </section><section><h3 id="legacy-assertion-mode">Legacy assertion mode
</h3> <p>Legacy assertion mode uses the <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Equality"><code>==</code> operator</a> in:</p> <ul> <li><a href="#assertdeepequalactual-expected-message"><code>assert.deepEqual()</code></a></li> <li><a href="#assertequalactual-expected-message"><code>assert.equal()</code></a></li> <li><a href="#assertnotdeepequalactual-expected-message"><code>assert.notDeepEqual()</code></a></li> <li><a href="#assertnotequalactual-expected-message"><code>assert.notEqual()</code></a></li> </ul> <p>To use legacy assertion mode:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');</pre> <p>Legacy assertion mode may have surprising results, especially when using <a href="#assertdeepequalactual-expected-message"><code>assert.deepEqual()</code></a>:</p> <pre data-language="js">// WARNING: This does not throw an AssertionError in legacy assertion mode!
assert.deepEqual(/a/gi, new Date());</pre> </section><section><h3 id="class-assertassertionerror">Class: assert.AssertionError<a class="srclink" href="https://github.com/nodejs/node/blob/6ac3a22f89ae43402d41166baf44266e60595fcb/lib/assert.js#L177">[src]</a>
</h3> <ul> <li>Extends: <a href="errors#class-error" class="type">&lt;errors.Error&gt;</a>
</li> </ul> <p>Indicates the failure of an assertion. All errors thrown by the <code>node:assert</code> module will be instances of the <code>AssertionError</code> class.</p> <h4 id="new-assertassertionerroroptions">
<code>new assert.AssertionError(options)</code>
</h4> <div class="api_metadata"> <span>Added in: v0.1.21</span> </div> <ul> <li>
<code>options</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object" class="type">&lt;Object&gt;</a> <ul> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> If provided, the error message is set to this value.</li> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a> The <code>actual</code> property on the error instance.</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a> The <code>expected</code> property on the error instance.</li> <li>
<code>operator</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> The <code>operator</code> property on the error instance.</li> <li>
<code>stackStartFn</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a> If provided, the generated stack trace omits frames before this function.</li> </ul> </li> </ul> <p>A subclass of <code>Error</code> that indicates the failure of an assertion.</p> <p>All instances contain the built-in <code>Error</code> properties (<code>message</code> and <code>name</code>) and:</p> <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a> Set to the <code>actual</code> argument for methods such as <a href="#assertstrictequalactual-expected-message"><code>assert.strictEqual()</code></a>.</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a> Set to the <code>expected</code> value for methods such as <a href="#assertstrictequalactual-expected-message"><code>assert.strictEqual()</code></a>.</li> <li>
<code>generatedMessage</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type" class="type">&lt;boolean&gt;</a> Indicates if the message was auto-generated (<code>true</code>) or not.</li> <li>
<code>code</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> Value is always <code>ERR_ASSERTION</code> to show that the error is an assertion error.</li> <li>
<code>operator</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> Set to the passed in operator value.</li> </ul> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

// Generate an AssertionError to compare the error message later:
const { message } = new assert.AssertionError({
  actual: 1,
  expected: 2,
  operator: 'strictEqual'
});

// Verify error output:
try {
  assert.strictEqual(1, 2);
} catch (err) {
  assert(err instanceof assert.AssertionError);
  assert.strictEqual(err.message, message);
  assert.strictEqual(err.name, 'AssertionError');
  assert.strictEqual(err.actual, 1);
  assert.strictEqual(err.expected, 2);
  assert.strictEqual(err.code, 'ERR_ASSERTION');
  assert.strictEqual(err.operator, 'strictEqual');
  assert.strictEqual(err.generatedMessage, true);
}</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

// Generate an AssertionError to compare the error message later:
const { message } = new assert.AssertionError({
  actual: 1,
  expected: 2,
  operator: 'strictEqual'
});

// Verify error output:
try {
  assert.strictEqual(1, 2);
} catch (err) {
  assert(err instanceof assert.AssertionError);
  assert.strictEqual(err.message, message);
  assert.strictEqual(err.name, 'AssertionError');
  assert.strictEqual(err.actual, 1);
  assert.strictEqual(err.expected, 2);
  assert.strictEqual(err.code, 'ERR_ASSERTION');
  assert.strictEqual(err.operator, 'strictEqual');
  assert.strictEqual(err.generatedMessage, true);
}</pre> </section><section><h3 id="class-assertcalltracker">Class: <code>assert.CallTracker</code>
</h3> <div class="api_metadata"> <span>Added in: v14.2.0, v12.19.0</span> </div> 
<div class="api_stability api_stability_1">
<a href="https://nodejs.org/api/documentation.html#stability-index">Stability: 1</a> - Experimental</div>
 <p>This feature is currently experimental and behavior might still change.</p> <h4 id="new-assertcalltracker">
<code>new assert.CallTracker()</code>
</h4> <div class="api_metadata"> <span>Added in: v14.2.0, v12.19.0</span> </div> <p>Creates a new <a href="#class-assertcalltracker"><code>CallTracker</code></a> object which can be used to track if functions were called a specific number of times. The <code>tracker.verify()</code> must be called for the verification to take place. The usual pattern would be to call it in a <a href="process#event-exit"><code>process.on('exit')</code></a> handler.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';
import process from 'node:process';

const tracker = new assert.CallTracker();

function func() {}

// callsfunc() must be called exactly 1 time before tracker.verify().
const callsfunc = tracker.calls(func, 1);

callsfunc();

// Calls tracker.verify() and verifies if all tracker.calls() functions have
// been called exact times.
process.on('exit', () =&gt; {
  tracker.verify();
});</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

const tracker = new assert.CallTracker();

function func() {}

// callsfunc() must be called exactly 1 time before tracker.verify().
const callsfunc = tracker.calls(func, 1);

callsfunc();

// Calls tracker.verify() and verifies if all tracker.calls() functions have
// been called exact times.
process.on('exit', () =&gt; {
  tracker.verify();
});</pre> <h4 id="trackercallsfn-exact">
<code>tracker.calls([fn][, exact])</code>
</h4> <div class="api_metadata"> <span>Added in: v14.2.0, v12.19.0</span> </div> <ul> <li>
<code>fn</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a> <strong>Default:</strong> A no-op function.</li> <li>
<code>exact</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type" class="type">&lt;number&gt;</a> <strong>Default:</strong> <code>1</code>.</li> <li>Returns: <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a> that wraps <code>fn</code>.</li> </ul> <p>The wrapper function is expected to be called exactly <code>exact</code> times. If the function has not been called exactly <code>exact</code> times when <a href="#trackerverify"><code>tracker.verify()</code></a> is called, then <a href="#trackerverify"><code>tracker.verify()</code></a> will throw an error.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

// Creates call tracker.
const tracker = new assert.CallTracker();

function func() {}

// Returns a function that wraps func() that must be called exact times
// before tracker.verify().
const callsfunc = tracker.calls(func);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

// Creates call tracker.
const tracker = new assert.CallTracker();

function func() {}

// Returns a function that wraps func() that must be called exact times
// before tracker.verify().
const callsfunc = tracker.calls(func);</pre> <h4 id="trackergetcallsfn">
<code>tracker.getCalls(fn)</code>
</h4> <div class="api_metadata"> <span>Added in: v18.8.0, v16.18.0</span> </div> <ul> <li> <p><code>fn</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a>.</p> </li> <li> <p>Returns: <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array" class="type">&lt;Array&gt;</a> with all the calls to a tracked function.</p> </li> <li> <p>Object <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object" class="type">&lt;Object&gt;</a></p> <ul> <li>
<code>thisArg</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object" class="type">&lt;Object&gt;</a>
</li> <li>
<code>arguments</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array" class="type">&lt;Array&gt;</a> the arguments passed to the tracked function</li> </ul> </li> </ul> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

const tracker = new assert.CallTracker();

function func() {}
const callsfunc = tracker.calls(func);
callsfunc(1, 2, 3);

assert.deepStrictEqual(tracker.getCalls(callsfunc),
                       [{ thisArg: this, arguments: [1, 2, 3 ] }]);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

// Creates call tracker.
const tracker = new assert.CallTracker();

function func() {}
const callsfunc = tracker.calls(func);
callsfunc(1, 2, 3);

assert.deepStrictEqual(tracker.getCalls(callsfunc),
                       [{ thisArg: this, arguments: [1, 2, 3 ] }]);</pre> <h4 id="trackerreport">
<code>tracker.report()</code>
</h4> <div class="api_metadata"> <span>Added in: v14.2.0, v12.19.0</span> </div> <ul> <li>Returns: <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array" class="type">&lt;Array&gt;</a> of objects containing information about the wrapper functions returned by <a href="#trackercallsfn-exact"><code>tracker.calls()</code></a>.</li> <li>Object <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object" class="type">&lt;Object&gt;</a> <ul> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a>
</li> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type" class="type">&lt;number&gt;</a> The actual number of times the function was called.</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type" class="type">&lt;number&gt;</a> The number of times the function was expected to be called.</li> <li>
<code>operator</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> The name of the function that is wrapped.</li> <li>
<code>stack</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object" class="type">&lt;Object&gt;</a> A stack trace of the function.</li> </ul> </li> </ul> <p>The arrays contains information about the expected and actual number of calls of the functions that have not been called the expected number of times.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

// Creates call tracker.
const tracker = new assert.CallTracker();

function func() {}

// Returns a function that wraps func() that must be called exact times
// before tracker.verify().
const callsfunc = tracker.calls(func, 2);

// Returns an array containing information on callsfunc()
tracker.report();
// [
//  {
//    message: 'Expected the func function to be executed 2 time(s) but was
//    executed 0 time(s).',
//    actual: 0,
//    expected: 2,
//    operator: 'func',
//    stack: stack trace
//  }
// ]</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

// Creates call tracker.
const tracker = new assert.CallTracker();

function func() {}

// Returns a function that wraps func() that must be called exact times
// before tracker.verify().
const callsfunc = tracker.calls(func, 2);

// Returns an array containing information on callsfunc()
tracker.report();
// [
//  {
//    message: 'Expected the func function to be executed 2 time(s) but was
//    executed 0 time(s).',
//    actual: 0,
//    expected: 2,
//    operator: 'func',
//    stack: stack trace
//  }
// ]</pre> <h4 id="trackerresetfn">
<code>tracker.reset([fn])</code>
</h4> <div class="api_metadata"> <span>Added in: v18.8.0, v16.18.0</span> </div> <ul> <li>
<code>fn</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a> a tracked function to reset.</li> </ul> <p>reset calls of the call tracker. if a tracked function is passed as an argument, the calls will be reset for it. if no arguments are passed, all tracked functions will be reset</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

const tracker = new assert.CallTracker();

function func() {}
const callsfunc = tracker.calls(func);

callsfunc();
// Tracker was called once
tracker.getCalls(callsfunc).length === 1;

tracker.reset(callsfunc);
tracker.getCalls(callsfunc).length === 0;</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

function func() {}
const callsfunc = tracker.calls(func);

callsfunc();
// Tracker was called once
tracker.getCalls(callsfunc).length === 1;

tracker.reset(callsfunc);
tracker.getCalls(callsfunc).length === 0;</pre> <h4 id="trackerverify">
<code>tracker.verify()</code>
</h4> <div class="api_metadata"> <span>Added in: v14.2.0, v12.19.0</span> </div> <p>Iterates through the list of functions passed to <a href="#trackercallsfn-exact"><code>tracker.calls()</code></a> and will throw an error for functions that have not been called the expected number of times.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

// Creates call tracker.
const tracker = new assert.CallTracker();

function func() {}

// Returns a function that wraps func() that must be called exact times
// before tracker.verify().
const callsfunc = tracker.calls(func, 2);

callsfunc();

// Will throw an error since callsfunc() was only called once.
tracker.verify();</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

// Creates call tracker.
const tracker = new assert.CallTracker();

function func() {}

// Returns a function that wraps func() that must be called exact times
// before tracker.verify().
const callsfunc = tracker.calls(func, 2);

callsfunc();

// Will throw an error since callsfunc() was only called once.
tracker.verify();</pre> </section><section><h3 id="assertvalue-message">
<code>assert(value[, message])</code>
</h3> <div class="api_metadata"> <span>Added in: v0.5.9</span> </div> <ul> <li>
<code>value</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a> The input that is checked for being truthy.</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p>An alias of <a href="#assertokvalue-message"><code>assert.ok()</code></a>.</p> </section><section><h3 id="assertdeepequalactual-expected-message">
<code>assert.deepEqual(actual, expected[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v18.0.0</td> <td><p>Regular expressions lastIndex property is now compared as well.</p></td>
</tr> <tr>
<td>v16.0.0, v14.18.0</td> <td><p>In Legacy assertion mode, changed status from Deprecated to Legacy.</p></td>
</tr> <tr>
<td>v14.0.0</td> <td><p>NaN is now treated as being identical if both sides are NaN.</p></td>
</tr> <tr>
<td>v12.0.0</td> <td><p>The type tags are now properly compared and there are a couple minor comparison adjustments to make the check less surprising.</p></td>
</tr> <tr>
<td>v9.0.0</td> <td><p>The <code>Error</code> names and messages are now properly compared.</p></td>
</tr> <tr>
<td>v8.0.0</td> <td><p>The <code>Set</code> and <code>Map</code> content is also compared.</p></td>
</tr> <tr>
<td>v6.4.0, v4.7.1</td> <td><p>Typed array slices are handled correctly now.</p></td>
</tr> <tr>
<td>v6.1.0, v4.5.0</td> <td><p>Objects with circular references can be used as inputs now.</p></td>
</tr> <tr>
<td>v5.10.1, v4.4.3</td> <td><p>Handle non-<code>Uint8Array</code> typed arrays correctly.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p><strong>Strict assertion mode</strong></p> <p>An alias of <a href="#assertdeepstrictequalactual-expected-message"><code>assert.deepStrictEqual()</code></a>.</p> <p><strong>Legacy assertion mode</strong></p> 
<div class="api_stability api_stability_3">
<a href="https://nodejs.org/api/documentation.html#stability-index">Stability: 3</a> - Legacy: Use <a href="#assertdeepstrictequalactual-expected-message"><code>assert.deepStrictEqual()</code></a> instead.</div>
 <p>Tests for deep equality between the <code>actual</code> and <code>expected</code> parameters. Consider using <a href="#assertdeepstrictequalactual-expected-message"><code>assert.deepStrictEqual()</code></a> instead. <a href="#assertdeepequalactual-expected-message"><code>assert.deepEqual()</code></a> can have surprising results.</p> <p><em>Deep equality</em> means that the enumerable "own" properties of child objects are also recursively evaluated by the following rules.</p> <h4 id="comparison-details">Comparison details
</h4> <ul> <li>Primitive values are compared with the <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Equality"><code>==</code> operator</a>, with the exception of <code>NaN</code>. It is treated as being identical in case both sides are <code>NaN</code>.</li> <li>
<a href="https://tc39.github.io/ecma262/#sec-object.prototype.tostring">Type tags</a> of objects should be the same.</li> <li>Only <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties">enumerable "own" properties</a> are considered.</li> <li>
<a href="errors#class-error"><code>Error</code></a> names and messages are always compared, even if these are not enumerable properties.</li> <li>
<a href="https://developer.mozilla.org/en-US/docs/Glossary/Primitive#Primitive_wrapper_objects_in_JavaScript">Object wrappers</a> are compared both as objects and unwrapped values.</li> <li>
<code>Object</code> properties are compared unordered.</li> <li>
<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map"><code>Map</code></a> keys and <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set"><code>Set</code></a> items are compared unordered.</li> <li>Recursion stops when both sides differ or both sides encounter a circular reference.</li> <li>Implementation does not test the <a href="https://tc39.github.io/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots"><code>[[Prototype]]</code></a> of objects.</li> <li>
<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol"><code>Symbol</code></a> properties are not compared.</li> <li>
<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap"><code>WeakMap</code></a> and <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakSet"><code>WeakSet</code></a> comparison does not rely on their values.</li> <li>
<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions"><code>RegExp</code></a> lastIndex, flags, and source are always compared, even if these are not enumerable properties.</li> </ul> <p>The following example does not throw an <a href="#class-assertassertionerror"><code>AssertionError</code></a> because the primitives are compared using the <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Equality"><code>==</code> operator</a>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';
// WARNING: This does not throw an AssertionError!

assert.deepEqual('+00000000', false);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');
// WARNING: This does not throw an AssertionError!

assert.deepEqual('+00000000', false);</pre> <p>"Deep" equality means that the enumerable "own" properties of child objects are evaluated also:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

const obj1 = {
  a: {
    b: 1
  }
};
const obj2 = {
  a: {
    b: 2
  }
};
const obj3 = {
  a: {
    b: 1
  }
};
const obj4 = Object.create(obj1);

assert.deepEqual(obj1, obj1);
// OK

// Values of b are different:
assert.deepEqual(obj1, obj2);
// AssertionError: { a: { b: 1 } } deepEqual { a: { b: 2 } }

assert.deepEqual(obj1, obj3);
// OK

// Prototypes are ignored:
assert.deepEqual(obj1, obj4);
// AssertionError: { a: { b: 1 } } deepEqual {}</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

const obj1 = {
  a: {
    b: 1
  }
};
const obj2 = {
  a: {
    b: 2
  }
};
const obj3 = {
  a: {
    b: 1
  }
};
const obj4 = Object.create(obj1);

assert.deepEqual(obj1, obj1);
// OK

// Values of b are different:
assert.deepEqual(obj1, obj2);
// AssertionError: { a: { b: 1 } } deepEqual { a: { b: 2 } }

assert.deepEqual(obj1, obj3);
// OK

// Prototypes are ignored:
assert.deepEqual(obj1, obj4);
// AssertionError: { a: { b: 1 } } deepEqual {}</pre> <p>If the values are not equal, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <a href="#class-assertassertionerror"><code>AssertionError</code></a>.</p> </section><section><h3 id="assertdeepstrictequalactual-expected-message">
<code>assert.deepStrictEqual(actual, expected[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v18.0.0</td> <td><p>Regular expressions lastIndex property is now compared as well.</p></td>
</tr> <tr>
<td>v9.0.0</td> <td><p>Enumerable symbol properties are now compared.</p></td>
</tr> <tr>
<td>v9.0.0</td> <td><p>The <code>NaN</code> is now compared using the <a href="https://tc39.github.io/ecma262/#sec-samevaluezero">SameValueZero</a> comparison.</p></td>
</tr> <tr>
<td>v8.5.0</td> <td><p>The <code>Error</code> names and messages are now properly compared.</p></td>
</tr> <tr>
<td>v8.0.0</td> <td><p>The <code>Set</code> and <code>Map</code> content is also compared.</p></td>
</tr> <tr>
<td>v6.1.0</td> <td><p>Objects with circular references can be used as inputs now.</p></td>
</tr> <tr>
<td>v6.4.0, v4.7.1</td> <td><p>Typed array slices are handled correctly now.</p></td>
</tr> <tr>
<td>v5.10.1, v4.4.3</td> <td><p>Handle non-<code>Uint8Array</code> typed arrays correctly.</p></td>
</tr> <tr>
<td>v1.2.0</td> <td><p><span>Added in: v1.2.0</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p>Tests for deep equality between the <code>actual</code> and <code>expected</code> parameters. "Deep" equality means that the enumerable "own" properties of child objects are recursively evaluated also by the following rules.</p> <h4 id="comparison-details_1">Comparison details
</h4> <ul> <li>Primitive values are compared using <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is"><code>Object.is()</code></a>.</li> <li>
<a href="https://tc39.github.io/ecma262/#sec-object.prototype.tostring">Type tags</a> of objects should be the same.</li> <li>
<a href="https://tc39.github.io/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots"><code>[[Prototype]]</code></a> of objects are compared using the <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality"><code>===</code> operator</a>.</li> <li>Only <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties">enumerable "own" properties</a> are considered.</li> <li>
<a href="errors#class-error"><code>Error</code></a> names and messages are always compared, even if these are not enumerable properties.</li> <li>Enumerable own <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol"><code>Symbol</code></a> properties are compared as well.</li> <li>
<a href="https://developer.mozilla.org/en-US/docs/Glossary/Primitive#Primitive_wrapper_objects_in_JavaScript">Object wrappers</a> are compared both as objects and unwrapped values.</li> <li>
<code>Object</code> properties are compared unordered.</li> <li>
<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map"><code>Map</code></a> keys and <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set"><code>Set</code></a> items are compared unordered.</li> <li>Recursion stops when both sides differ or both sides encounter a circular reference.</li> <li>
<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap"><code>WeakMap</code></a> and <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakSet"><code>WeakSet</code></a> comparison does not rely on their values. See below for further details.</li> <li>
<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions"><code>RegExp</code></a> lastIndex, flags, and source are always compared, even if these are not enumerable properties.</li> </ul> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

// This fails because 1 !== '1'.
assert.deepStrictEqual({ a: 1 }, { a: '1' });
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
//   {
// +   a: 1
// -   a: '1'
//   }

// The following objects don't have own properties
const date = new Date();
const object = {};
const fakeDate = {};
Object.setPrototypeOf(fakeDate, Date.prototype);

// Different [[Prototype]]:
assert.deepStrictEqual(object, fakeDate);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
// + {}
// - Date {}

// Different type tags:
assert.deepStrictEqual(date, fakeDate);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
// + 2018-04-26T00:49:08.604Z
// - Date {}

assert.deepStrictEqual(NaN, NaN);
// OK because Object.is(NaN, NaN) is true.

// Different unwrapped numbers:
assert.deepStrictEqual(new Number(1), new Number(2));
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
// + [Number: 1]
// - [Number: 2]

assert.deepStrictEqual(new String('foo'), Object('foo'));
// OK because the object and the string are identical when unwrapped.

assert.deepStrictEqual(-0, -0);
// OK

// Different zeros:
assert.deepStrictEqual(0, -0);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
// + 0
// - -0

const symbol1 = Symbol();
const symbol2 = Symbol();
assert.deepStrictEqual({ [symbol1]: 1 }, { [symbol1]: 1 });
// OK, because it is the same symbol on both objects.

assert.deepStrictEqual({ [symbol1]: 1 }, { [symbol2]: 1 });
// AssertionError [ERR_ASSERTION]: Inputs identical but not reference equal:
//
// {
//   [Symbol()]: 1
// }

const weakMap1 = new WeakMap();
const weakMap2 = new WeakMap([[{}, {}]]);
const weakMap3 = new WeakMap();
weakMap3.unequal = true;

assert.deepStrictEqual(weakMap1, weakMap2);
// OK, because it is impossible to compare the entries

// Fails because weakMap3 has a property that weakMap1 does not contain:
assert.deepStrictEqual(weakMap1, weakMap3);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
//   WeakMap {
// +   [items unknown]
// -   [items unknown],
// -   unequal: true
//   }</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

// This fails because 1 !== '1'.
assert.deepStrictEqual({ a: 1 }, { a: '1' });
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
//   {
// +   a: 1
// -   a: '1'
//   }

// The following objects don't have own properties
const date = new Date();
const object = {};
const fakeDate = {};
Object.setPrototypeOf(fakeDate, Date.prototype);

// Different [[Prototype]]:
assert.deepStrictEqual(object, fakeDate);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
// + {}
// - Date {}

// Different type tags:
assert.deepStrictEqual(date, fakeDate);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
// + 2018-04-26T00:49:08.604Z
// - Date {}

assert.deepStrictEqual(NaN, NaN);
// OK because Object.is(NaN, NaN) is true.

// Different unwrapped numbers:
assert.deepStrictEqual(new Number(1), new Number(2));
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
// + [Number: 1]
// - [Number: 2]

assert.deepStrictEqual(new String('foo'), Object('foo'));
// OK because the object and the string are identical when unwrapped.

assert.deepStrictEqual(-0, -0);
// OK

// Different zeros:
assert.deepStrictEqual(0, -0);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
// + 0
// - -0

const symbol1 = Symbol();
const symbol2 = Symbol();
assert.deepStrictEqual({ [symbol1]: 1 }, { [symbol1]: 1 });
// OK, because it is the same symbol on both objects.

assert.deepStrictEqual({ [symbol1]: 1 }, { [symbol2]: 1 });
// AssertionError [ERR_ASSERTION]: Inputs identical but not reference equal:
//
// {
//   [Symbol()]: 1
// }

const weakMap1 = new WeakMap();
const weakMap2 = new WeakMap([[{}, {}]]);
const weakMap3 = new WeakMap();
weakMap3.unequal = true;

assert.deepStrictEqual(weakMap1, weakMap2);
// OK, because it is impossible to compare the entries

// Fails because weakMap3 has a property that weakMap1 does not contain:
assert.deepStrictEqual(weakMap1, weakMap3);
// AssertionError: Expected inputs to be strictly deep-equal:
// + actual - expected
//
//   WeakMap {
// +   [items unknown]
// -   [items unknown],
// -   unequal: true
//   }</pre> <p>If the values are not equal, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <code>AssertionError</code>.</p> </section><section><h3 id="assertdoesnotmatchstring-regexp-message">
<code>assert.doesNotMatch(string, regexp[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v16.0.0</td> <td><p>This API is no longer experimental.</p></td>
</tr> <tr>
<td>v13.6.0, v12.16.0</td> <td><p><span>Added in: v13.6.0, v12.16.0</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>string</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a>
</li> <li>
<code>regexp</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp" class="type">&lt;RegExp&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p>Expects the <code>string</code> input not to match the regular expression.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.doesNotMatch('I will fail', /fail/);
// AssertionError [ERR_ASSERTION]: The input was expected to not match the ...

assert.doesNotMatch(123, /pass/);
// AssertionError [ERR_ASSERTION]: The "string" argument must be of type string.

assert.doesNotMatch('I will pass', /different/);
// OK</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.doesNotMatch('I will fail', /fail/);
// AssertionError [ERR_ASSERTION]: The input was expected to not match the ...

assert.doesNotMatch(123, /pass/);
// AssertionError [ERR_ASSERTION]: The "string" argument must be of type string.

assert.doesNotMatch('I will pass', /different/);
// OK</pre> <p>If the values do match, or if the <code>string</code> argument is of another type than <code>string</code>, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <a href="#class-assertassertionerror"><code>AssertionError</code></a>.</p> </section><section><h3 id="assertdoesnotrejectasyncfn-error-message">
<code>assert.doesNotReject(asyncFn[, error][, message])</code>
</h3> <div class="api_metadata"> <span>Added in: v10.0.0</span> </div> <ul> <li>
<code>asyncFn</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise" class="type">&lt;Promise&gt;</a>
</li> <li>
<code>error</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp" class="type">&lt;RegExp&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a>
</li> </ul> <p>Awaits the <code>asyncFn</code> promise or, if <code>asyncFn</code> is a function, immediately calls the function and awaits the returned promise to complete. It will then check that the promise is not rejected.</p> <p>If <code>asyncFn</code> is a function and it throws an error synchronously, <code>assert.doesNotReject()</code> will return a rejected <code>Promise</code> with that error. If the function does not return a promise, <code>assert.doesNotReject()</code> will return a rejected <code>Promise</code> with an <a href="errors#err_invalid_return_value"><code>ERR_INVALID_RETURN_VALUE</code></a> error. In both cases the error handler is skipped.</p> <p>Using <code>assert.doesNotReject()</code> is actually not useful because there is little benefit in catching a rejection and then rejecting it again. Instead, consider adding a comment next to the specific code path that should not reject and keep error messages as expressive as possible.</p> <p>If specified, <code>error</code> can be a <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes"><code>Class</code></a>, <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions"><code>RegExp</code></a>, or a validation function. See <a href="#assertthrowsfn-error-message"><code>assert.throws()</code></a> for more details.</p> <p>Besides the async nature to await the completion behaves identically to <a href="#assertdoesnotthrowfn-error-message"><code>assert.doesNotThrow()</code></a>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

await assert.doesNotReject(
  async () =&gt; {
    throw new TypeError('Wrong value');
  },
  SyntaxError
);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

(async () =&gt; {
  await assert.doesNotReject(
    async () =&gt; {
      throw new TypeError('Wrong value');
    },
    SyntaxError
  );
})();</pre> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.doesNotReject(Promise.reject(new TypeError('Wrong value')))
  .then(() =&gt; {
    // ...
  });</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.doesNotReject(Promise.reject(new TypeError('Wrong value')))
  .then(() =&gt; {
    // ...
  });</pre> </section><section><h3 id="assertdoesnotthrowfn-error-message">
<code>assert.doesNotThrow(fn[, error][, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v5.11.0, v4.4.5</td> <td><p>The <code>message</code> parameter is respected now.</p></td>
</tr> <tr>
<td>v4.2.0</td> <td><p>The <code>error</code> parameter can now be an arrow function.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>fn</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a>
</li> <li>
<code>error</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp" class="type">&lt;RegExp&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a>
</li> </ul> <p>Asserts that the function <code>fn</code> does not throw an error.</p> <p>Using <code>assert.doesNotThrow()</code> is actually not useful because there is no benefit in catching an error and then rethrowing it. Instead, consider adding a comment next to the specific code path that should not throw and keep error messages as expressive as possible.</p> <p>When <code>assert.doesNotThrow()</code> is called, it will immediately call the <code>fn</code> function.</p> <p>If an error is thrown and it is the same type as that specified by the <code>error</code> parameter, then an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown. If the error is of a different type, or if the <code>error</code> parameter is undefined, the error is propagated back to the caller.</p> <p>If specified, <code>error</code> can be a <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes"><code>Class</code></a>, <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions"><code>RegExp</code></a>, or a validation function. See <a href="#assertthrowsfn-error-message"><code>assert.throws()</code></a> for more details.</p> <p>The following, for instance, will throw the <a href="errors#class-typeerror"><code>TypeError</code></a> because there is no matching error type in the assertion:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.doesNotThrow(
  () =&gt; {
    throw new TypeError('Wrong value');
  },
  SyntaxError
);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.doesNotThrow(
  () =&gt; {
    throw new TypeError('Wrong value');
  },
  SyntaxError
);</pre> <p>However, the following will result in an <a href="#class-assertassertionerror"><code>AssertionError</code></a> with the message 'Got unwanted exception...':</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.doesNotThrow(
  () =&gt; {
    throw new TypeError('Wrong value');
  },
  TypeError
);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.doesNotThrow(
  () =&gt; {
    throw new TypeError('Wrong value');
  },
  TypeError
);</pre> <p>If an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown and a value is provided for the <code>message</code> parameter, the value of <code>message</code> will be appended to the <a href="#class-assertassertionerror"><code>AssertionError</code></a> message:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.doesNotThrow(
  () =&gt; {
    throw new TypeError('Wrong value');
  },
  /Wrong value/,
  'Whoops'
);
// Throws: AssertionError: Got unwanted exception: Whoops</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.doesNotThrow(
  () =&gt; {
    throw new TypeError('Wrong value');
  },
  /Wrong value/,
  'Whoops'
);
// Throws: AssertionError: Got unwanted exception: Whoops</pre> </section><section><h3 id="assertequalactual-expected-message">
<code>assert.equal(actual, expected[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v16.0.0, v14.18.0</td> <td><p>In Legacy assertion mode, changed status from Deprecated to Legacy.</p></td>
</tr> <tr>
<td>v14.0.0</td> <td><p>NaN is now treated as being identical if both sides are NaN.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p><strong>Strict assertion mode</strong></p> <p>An alias of <a href="#assertstrictequalactual-expected-message"><code>assert.strictEqual()</code></a>.</p> <p><strong>Legacy assertion mode</strong></p> 
<div class="api_stability api_stability_3">
<a href="https://nodejs.org/api/documentation.html#stability-index">Stability: 3</a> - Legacy: Use <a href="#assertstrictequalactual-expected-message"><code>assert.strictEqual()</code></a> instead.</div>
 <p>Tests shallow, coercive equality between the <code>actual</code> and <code>expected</code> parameters using the <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Equality"><code>==</code> operator</a>. <code>NaN</code> is specially handled and treated as being identical if both sides are <code>NaN</code>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

assert.equal(1, 1);
// OK, 1 == 1
assert.equal(1, '1');
// OK, 1 == '1'
assert.equal(NaN, NaN);
// OK

assert.equal(1, 2);
// AssertionError: 1 == 2
assert.equal({ a: { b: 1 } }, { a: { b: 1 } });
// AssertionError: { a: { b: 1 } } == { a: { b: 1 } }</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

assert.equal(1, 1);
// OK, 1 == 1
assert.equal(1, '1');
// OK, 1 == '1'
assert.equal(NaN, NaN);
// OK

assert.equal(1, 2);
// AssertionError: 1 == 2
assert.equal({ a: { b: 1 } }, { a: { b: 1 } });
// AssertionError: { a: { b: 1 } } == { a: { b: 1 } }</pre> <p>If the values are not equal, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <code>AssertionError</code>.</p> </section><section><h3 id="assertfailmessage">
<code>assert.fail([message])</code>
</h3> <div class="api_metadata"> <span>Added in: v0.1.21</span> </div> <ul> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a> <strong>Default:</strong> <code>'Failed'</code>
</li> </ul> <p>Throws an <a href="#class-assertassertionerror"><code>AssertionError</code></a> with the provided error message or a default error message. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <a href="#class-assertassertionerror"><code>AssertionError</code></a>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.fail();
// AssertionError [ERR_ASSERTION]: Failed

assert.fail('boom');
// AssertionError [ERR_ASSERTION]: boom

assert.fail(new TypeError('need array'));
// TypeError: need array</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.fail();
// AssertionError [ERR_ASSERTION]: Failed

assert.fail('boom');
// AssertionError [ERR_ASSERTION]: boom

assert.fail(new TypeError('need array'));
// TypeError: need array</pre> <p>Using <code>assert.fail()</code> with more than two arguments is possible but deprecated. See below for further details.</p> </section><section><h3 id="assertfailactual-expected-message-operator-stackstartfn">
<code>assert.fail(actual, expected[, message[, operator[, stackStartFn]]])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v10.0.0</td> <td><p>Calling <code>assert.fail()</code> with more than one argument is deprecated and emits a warning.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> 
<div class="api_stability api_stability_0">
<a href="https://nodejs.org/api/documentation.html#stability-index">Stability: 0</a> - Deprecated: Use <code>assert.fail([message])</code> or other assert functions instead.</div>
 <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> <li>
<code>operator</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> <strong>Default:</strong> <code>'!='</code>
</li> <li>
<code>stackStartFn</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a> <strong>Default:</strong> <code>assert.fail</code>
</li> </ul> <p>If <code>message</code> is falsy, the error message is set as the values of <code>actual</code> and <code>expected</code> separated by the provided <code>operator</code>. If just the two <code>actual</code> and <code>expected</code> arguments are provided, <code>operator</code> will default to <code>'!='</code>. If <code>message</code> is provided as third argument it will be used as the error message and the other arguments will be stored as properties on the thrown object. If <code>stackStartFn</code> is provided, all stack frames above that function will be removed from stacktrace (see <a href="errors#errorcapturestacktracetargetobject-constructoropt"><code>Error.captureStackTrace</code></a>). If no arguments are given, the default message <code>Failed</code> will be used.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.fail('a', 'b');
// AssertionError [ERR_ASSERTION]: 'a' != 'b'

assert.fail(1, 2, undefined, '&gt;');
// AssertionError [ERR_ASSERTION]: 1 &gt; 2

assert.fail(1, 2, 'fail');
// AssertionError [ERR_ASSERTION]: fail

assert.fail(1, 2, 'whoops', '&gt;');
// AssertionError [ERR_ASSERTION]: whoops

assert.fail(1, 2, new TypeError('need array'));
// TypeError: need array</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.fail('a', 'b');
// AssertionError [ERR_ASSERTION]: 'a' != 'b'

assert.fail(1, 2, undefined, '&gt;');
// AssertionError [ERR_ASSERTION]: 1 &gt; 2

assert.fail(1, 2, 'fail');
// AssertionError [ERR_ASSERTION]: fail

assert.fail(1, 2, 'whoops', '&gt;');
// AssertionError [ERR_ASSERTION]: whoops

assert.fail(1, 2, new TypeError('need array'));
// TypeError: need array</pre> <p>In the last three cases <code>actual</code>, <code>expected</code>, and <code>operator</code> have no influence on the error message.</p> <p>Example use of <code>stackStartFn</code> for truncating the exception's stacktrace:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

function suppressFrame() {
  assert.fail('a', 'b', undefined, '!==', suppressFrame);
}
suppressFrame();
// AssertionError [ERR_ASSERTION]: 'a' !== 'b'
//     at repl:1:1
//     at ContextifyScript.Script.runInThisContext (vm.js:44:33)
//     ...</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

function suppressFrame() {
  assert.fail('a', 'b', undefined, '!==', suppressFrame);
}
suppressFrame();
// AssertionError [ERR_ASSERTION]: 'a' !== 'b'
//     at repl:1:1
//     at ContextifyScript.Script.runInThisContext (vm.js:44:33)
//     ...</pre> </section><section><h3 id="assertiferrorvalue">
<code>assert.ifError(value)</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v10.0.0</td> <td><p>Instead of throwing the original error it is now wrapped into an [<code>AssertionError</code>][] that contains the full stack trace.</p></td>
</tr> <tr>
<td>v10.0.0</td> <td><p>Value may now only be <code>undefined</code> or <code>null</code>. Before all falsy values were handled the same as <code>null</code> and did not throw.</p></td>
</tr> <tr>
<td>v0.1.97</td> <td><p><span>Added in: v0.1.97</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>value</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> </ul> <p>Throws <code>value</code> if <code>value</code> is not <code>undefined</code> or <code>null</code>. This is useful when testing the <code>error</code> argument in callbacks. The stack trace contains all frames from the error passed to <code>ifError()</code> including the potential new frames for <code>ifError()</code> itself.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.ifError(null);
// OK
assert.ifError(0);
// AssertionError [ERR_ASSERTION]: ifError got unwanted exception: 0
assert.ifError('error');
// AssertionError [ERR_ASSERTION]: ifError got unwanted exception: 'error'
assert.ifError(new Error());
// AssertionError [ERR_ASSERTION]: ifError got unwanted exception: Error

// Create some random error frames.
let err;
(function errorFrame() {
  err = new Error('test error');
})();

(function ifErrorFrame() {
  assert.ifError(err);
})();
// AssertionError [ERR_ASSERTION]: ifError got unwanted exception: test error
//     at ifErrorFrame
//     at errorFrame</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.ifError(null);
// OK
assert.ifError(0);
// AssertionError [ERR_ASSERTION]: ifError got unwanted exception: 0
assert.ifError('error');
// AssertionError [ERR_ASSERTION]: ifError got unwanted exception: 'error'
assert.ifError(new Error());
// AssertionError [ERR_ASSERTION]: ifError got unwanted exception: Error

// Create some random error frames.
let err;
(function errorFrame() {
  err = new Error('test error');
})();

(function ifErrorFrame() {
  assert.ifError(err);
})();
// AssertionError [ERR_ASSERTION]: ifError got unwanted exception: test error
//     at ifErrorFrame
//     at errorFrame</pre> </section><section><h3 id="assertmatchstring-regexp-message">
<code>assert.match(string, regexp[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v16.0.0</td> <td><p>This API is no longer experimental.</p></td>
</tr> <tr>
<td>v13.6.0, v12.16.0</td> <td><p><span>Added in: v13.6.0, v12.16.0</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>string</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a>
</li> <li>
<code>regexp</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp" class="type">&lt;RegExp&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p>Expects the <code>string</code> input to match the regular expression.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.match('I will fail', /pass/);
// AssertionError [ERR_ASSERTION]: The input did not match the regular ...

assert.match(123, /pass/);
// AssertionError [ERR_ASSERTION]: The "string" argument must be of type string.

assert.match('I will pass', /pass/);
// OK</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.match('I will fail', /pass/);
// AssertionError [ERR_ASSERTION]: The input did not match the regular ...

assert.match(123, /pass/);
// AssertionError [ERR_ASSERTION]: The "string" argument must be of type string.

assert.match('I will pass', /pass/);
// OK</pre> <p>If the values do not match, or if the <code>string</code> argument is of another type than <code>string</code>, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <a href="#class-assertassertionerror"><code>AssertionError</code></a>.</p> </section><section><h3 id="assertnotdeepequalactual-expected-message">
<code>assert.notDeepEqual(actual, expected[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v16.0.0, v14.18.0</td> <td><p>In Legacy assertion mode, changed status from Deprecated to Legacy.</p></td>
</tr> <tr>
<td>v14.0.0</td> <td><p>NaN is now treated as being identical if both sides are NaN.</p></td>
</tr> <tr>
<td>v9.0.0</td> <td><p>The <code>Error</code> names and messages are now properly compared.</p></td>
</tr> <tr>
<td>v8.0.0</td> <td><p>The <code>Set</code> and <code>Map</code> content is also compared.</p></td>
</tr> <tr>
<td>v6.4.0, v4.7.1</td> <td><p>Typed array slices are handled correctly now.</p></td>
</tr> <tr>
<td>v6.1.0, v4.5.0</td> <td><p>Objects with circular references can be used as inputs now.</p></td>
</tr> <tr>
<td>v5.10.1, v4.4.3</td> <td><p>Handle non-<code>Uint8Array</code> typed arrays correctly.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p><strong>Strict assertion mode</strong></p> <p>An alias of <a href="#assertnotdeepstrictequalactual-expected-message"><code>assert.notDeepStrictEqual()</code></a>.</p> <p><strong>Legacy assertion mode</strong></p> 
<div class="api_stability api_stability_3">
<a href="https://nodejs.org/api/documentation.html#stability-index">Stability: 3</a> - Legacy: Use <a href="#assertnotdeepstrictequalactual-expected-message"><code>assert.notDeepStrictEqual()</code></a> instead.</div>
 <p>Tests for any deep inequality. Opposite of <a href="#assertdeepequalactual-expected-message"><code>assert.deepEqual()</code></a>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

const obj1 = {
  a: {
    b: 1
  }
};
const obj2 = {
  a: {
    b: 2
  }
};
const obj3 = {
  a: {
    b: 1
  }
};
const obj4 = Object.create(obj1);

assert.notDeepEqual(obj1, obj1);
// AssertionError: { a: { b: 1 } } notDeepEqual { a: { b: 1 } }

assert.notDeepEqual(obj1, obj2);
// OK

assert.notDeepEqual(obj1, obj3);
// AssertionError: { a: { b: 1 } } notDeepEqual { a: { b: 1 } }

assert.notDeepEqual(obj1, obj4);
// OK</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

const obj1 = {
  a: {
    b: 1
  }
};
const obj2 = {
  a: {
    b: 2
  }
};
const obj3 = {
  a: {
    b: 1
  }
};
const obj4 = Object.create(obj1);

assert.notDeepEqual(obj1, obj1);
// AssertionError: { a: { b: 1 } } notDeepEqual { a: { b: 1 } }

assert.notDeepEqual(obj1, obj2);
// OK

assert.notDeepEqual(obj1, obj3);
// AssertionError: { a: { b: 1 } } notDeepEqual { a: { b: 1 } }

assert.notDeepEqual(obj1, obj4);
// OK</pre> <p>If the values are deeply equal, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <code>AssertionError</code>.</p> </section><section><h3 id="assertnotdeepstrictequalactual-expected-message">
<code>assert.notDeepStrictEqual(actual, expected[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v9.0.0</td> <td><p>The <code>-0</code> and <code>+0</code> are not considered equal anymore.</p></td>
</tr> <tr>
<td>v9.0.0</td> <td><p>The <code>NaN</code> is now compared using the <a href="https://tc39.github.io/ecma262/#sec-samevaluezero">SameValueZero</a> comparison.</p></td>
</tr> <tr>
<td>v9.0.0</td> <td><p>The <code>Error</code> names and messages are now properly compared.</p></td>
</tr> <tr>
<td>v8.0.0</td> <td><p>The <code>Set</code> and <code>Map</code> content is also compared.</p></td>
</tr> <tr>
<td>v6.1.0</td> <td><p>Objects with circular references can be used as inputs now.</p></td>
</tr> <tr>
<td>v6.4.0, v4.7.1</td> <td><p>Typed array slices are handled correctly now.</p></td>
</tr> <tr>
<td>v5.10.1, v4.4.3</td> <td><p>Handle non-<code>Uint8Array</code> typed arrays correctly.</p></td>
</tr> <tr>
<td>v1.2.0</td> <td><p><span>Added in: v1.2.0</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p>Tests for deep strict inequality. Opposite of <a href="#assertdeepstrictequalactual-expected-message"><code>assert.deepStrictEqual()</code></a>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.notDeepStrictEqual({ a: 1 }, { a: '1' });
// OK</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.notDeepStrictEqual({ a: 1 }, { a: '1' });
// OK</pre> <p>If the values are deeply and strictly equal, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <a href="#class-assertassertionerror"><code>AssertionError</code></a>.</p> </section><section><h3 id="assertnotequalactual-expected-message">
<code>assert.notEqual(actual, expected[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v16.0.0, v14.18.0</td> <td><p>In Legacy assertion mode, changed status from Deprecated to Legacy.</p></td>
</tr> <tr>
<td>v14.0.0</td> <td><p>NaN is now treated as being identical if both sides are NaN.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p><strong>Strict assertion mode</strong></p> <p>An alias of <a href="#assertnotstrictequalactual-expected-message"><code>assert.notStrictEqual()</code></a>.</p> <p><strong>Legacy assertion mode</strong></p> 
<div class="api_stability api_stability_3">
<a href="https://nodejs.org/api/documentation.html#stability-index">Stability: 3</a> - Legacy: Use <a href="#assertnotstrictequalactual-expected-message"><code>assert.notStrictEqual()</code></a> instead.</div>
 <p>Tests shallow, coercive inequality with the <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Inequality"><code>!=</code> operator</a>. <code>NaN</code> is specially handled and treated as being identical if both sides are <code>NaN</code>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert';

assert.notEqual(1, 2);
// OK

assert.notEqual(1, 1);
// AssertionError: 1 != 1

assert.notEqual(1, '1');
// AssertionError: 1 != '1'</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

assert.notEqual(1, 2);
// OK

assert.notEqual(1, 1);
// AssertionError: 1 != 1

assert.notEqual(1, '1');
// AssertionError: 1 != '1'</pre> <p>If the values are equal, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <code>AssertionError</code>.</p> </section><section><h3 id="assertnotstrictequalactual-expected-message">
<code>assert.notStrictEqual(actual, expected[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v10.0.0</td> <td><p>Used comparison changed from Strict Equality to <code>Object.is()</code>.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p>Tests strict inequality between the <code>actual</code> and <code>expected</code> parameters as determined by <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is"><code>Object.is()</code></a>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.notStrictEqual(1, 2);
// OK

assert.notStrictEqual(1, 1);
// AssertionError [ERR_ASSERTION]: Expected "actual" to be strictly unequal to:
//
// 1

assert.notStrictEqual(1, '1');
// OK</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.notStrictEqual(1, 2);
// OK

assert.notStrictEqual(1, 1);
// AssertionError [ERR_ASSERTION]: Expected "actual" to be strictly unequal to:
//
// 1

assert.notStrictEqual(1, '1');
// OK</pre> <p>If the values are strictly equal, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <code>AssertionError</code>.</p> </section><section><h3 id="assertokvalue-message">
<code>assert.ok(value[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v10.0.0</td> <td><p>The <code>assert.ok()</code> (no arguments) will now use a predefined error message.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>value</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p>Tests if <code>value</code> is truthy. It is equivalent to <code>assert.equal(!!value, true, message)</code>.</p> <p>If <code>value</code> is not truthy, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is <code>undefined</code>, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <code>AssertionError</code>. If no arguments are passed in at all <code>message</code> will be set to the string: <code>'No value argument passed to `assert.ok()`'</code>.</p> <p>Be aware that in the <code>repl</code> the error message will be different to the one thrown in a file! See below for further details.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.ok(true);
// OK
assert.ok(1);
// OK

assert.ok();
// AssertionError: No value argument passed to `assert.ok()`

assert.ok(false, 'it\'s false');
// AssertionError: it's false

// In the repl:
assert.ok(typeof 123 === 'string');
// AssertionError: false == true

// In a file (e.g. test.js):
assert.ok(typeof 123 === 'string');
// AssertionError: The expression evaluated to a falsy value:
//
//   assert.ok(typeof 123 === 'string')

assert.ok(false);
// AssertionError: The expression evaluated to a falsy value:
//
//   assert.ok(false)

assert.ok(0);
// AssertionError: The expression evaluated to a falsy value:
//
//   assert.ok(0)</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.ok(true);
// OK
assert.ok(1);
// OK

assert.ok();
// AssertionError: No value argument passed to `assert.ok()`

assert.ok(false, 'it\'s false');
// AssertionError: it's false

// In the repl:
assert.ok(typeof 123 === 'string');
// AssertionError: false == true

// In a file (e.g. test.js):
assert.ok(typeof 123 === 'string');
// AssertionError: The expression evaluated to a falsy value:
//
//   assert.ok(typeof 123 === 'string')

assert.ok(false);
// AssertionError: The expression evaluated to a falsy value:
//
//   assert.ok(false)

assert.ok(0);
// AssertionError: The expression evaluated to a falsy value:
//
//   assert.ok(0)</pre> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

// Using `assert()` works the same:
assert(0);
// AssertionError: The expression evaluated to a falsy value:
//
//   assert(0)</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert');

// Using `assert()` works the same:
assert(0);
// AssertionError: The expression evaluated to a falsy value:
//
//   assert(0)</pre> </section><section><h3 id="assertrejectsasyncfn-error-message">
<code>assert.rejects(asyncFn[, error][, message])</code>
</h3> <div class="api_metadata"> <span>Added in: v10.0.0</span> </div> <ul> <li>
<code>asyncFn</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise" class="type">&lt;Promise&gt;</a>
</li> <li>
<code>error</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp" class="type">&lt;RegExp&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object" class="type">&lt;Object&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a>
</li> </ul> <p>Awaits the <code>asyncFn</code> promise or, if <code>asyncFn</code> is a function, immediately calls the function and awaits the returned promise to complete. It will then check that the promise is rejected.</p> <p>If <code>asyncFn</code> is a function and it throws an error synchronously, <code>assert.rejects()</code> will return a rejected <code>Promise</code> with that error. If the function does not return a promise, <code>assert.rejects()</code> will return a rejected <code>Promise</code> with an <a href="errors#err_invalid_return_value"><code>ERR_INVALID_RETURN_VALUE</code></a> error. In both cases the error handler is skipped.</p> <p>Besides the async nature to await the completion behaves identically to <a href="#assertthrowsfn-error-message"><code>assert.throws()</code></a>.</p> <p>If specified, <code>error</code> can be a <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes"><code>Class</code></a>, <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions"><code>RegExp</code></a>, a validation function, an object where each property will be tested for, or an instance of error where each property will be tested for including the non-enumerable <code>message</code> and <code>name</code> properties.</p> <p>If specified, <code>message</code> will be the message provided by the <a href="#class-assertassertionerror"><code>AssertionError</code></a> if the <code>asyncFn</code> fails to reject.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

await assert.rejects(
  async () =&gt; {
    throw new TypeError('Wrong value');
  },
  {
    name: 'TypeError',
    message: 'Wrong value'
  }
);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

(async () =&gt; {
  await assert.rejects(
    async () =&gt; {
      throw new TypeError('Wrong value');
    },
    {
      name: 'TypeError',
      message: 'Wrong value'
    }
  );
})();</pre> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

await assert.rejects(
  async () =&gt; {
    throw new TypeError('Wrong value');
  },
  (err) =&gt; {
    assert.strictEqual(err.name, 'TypeError');
    assert.strictEqual(err.message, 'Wrong value');
    return true;
  }
);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

(async () =&gt; {
  await assert.rejects(
    async () =&gt; {
      throw new TypeError('Wrong value');
    },
    (err) =&gt; {
      assert.strictEqual(err.name, 'TypeError');
      assert.strictEqual(err.message, 'Wrong value');
      return true;
    }
  );
})();</pre> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.rejects(
  Promise.reject(new Error('Wrong value')),
  Error
).then(() =&gt; {
  // ...
});</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.rejects(
  Promise.reject(new Error('Wrong value')),
  Error
).then(() =&gt; {
  // ...
});</pre> <p><code>error</code> cannot be a string. If a string is provided as the second argument, then <code>error</code> is assumed to be omitted and the string will be used for <code>message</code> instead. This can lead to easy-to-miss mistakes. Please read the example in <a href="#assertthrowsfn-error-message"><code>assert.throws()</code></a> carefully if using a string as the second argument gets considered.</p> </section><section><h3 id="assertsnapshotvalue-name">
<code>assert.snapshot(value, name)</code>
</h3> <div class="api_metadata"> <span>Added in: v18.8.0</span> </div> 
<div class="api_stability api_stability_1">
<a href="https://nodejs.org/api/documentation.html#stability-index">Stability: 1</a> - Experimental</div>
 <ul> <li>
<code>value</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a> the value to snapshot.</li> <li>
<code>name</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> the name of the snapshot.</li> <li>Returns: <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise" class="type">&lt;Promise&gt;</a>
</li> </ul> <p>Reads the <code>name</code> snapshot from a file and compares <code>value</code> to the snapshot. <code>value</code> is serialized with <a href="util#utilinspectobject-options"><code>util.inspect()</code></a>. If the value is not strictly equal to the snapshot, <code>assert.snapshot()</code> returns a rejected <code>Promise</code> with an <a href="#class-assertassertionerror"><code>AssertionError</code></a>.</p> <p>The snapshot filename uses the same basename as the application's main entrypoint with a <code>.snapshot</code> extension. If the snapshot file does not exist, it is created. The <a href="cli#--update-assert-snapshot"><code>--update-assert-snapshot</code></a> command line flag can be used to force the update of an existing snapshot.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

// Assuming that the application's main entrypoint is app.mjs, this reads the
// 'snapshotName' snapshot from app.snapshot and strictly compares its value
// to `util.inspect('value')`.
await assert.snapshot('value', 'snapshotName');</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

(async () =&gt; {
  // Assuming that the application's main entrypoint is app.js, this reads the
  // 'snapshotName' snapshot from app.snapshot and strictly compares its value
  // to `util.inspect('value')`.
  await assert.snapshot('value', 'snapshotName');
})();</pre> </section><section><h3 id="assertstrictequalactual-expected-message">
<code>assert.strictEqual(actual, expected[, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v10.0.0</td> <td><p>Used comparison changed from Strict Equality to <code>Object.is()</code>.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>actual</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>expected</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types" class="type">&lt;any&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> </ul> <p>Tests strict equality between the <code>actual</code> and <code>expected</code> parameters as determined by <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is"><code>Object.is()</code></a>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.strictEqual(1, 2);
// AssertionError [ERR_ASSERTION]: Expected inputs to be strictly equal:
//
// 1 !== 2

assert.strictEqual(1, 1);
// OK

assert.strictEqual('Hello foobar', 'Hello World!');
// AssertionError [ERR_ASSERTION]: Expected inputs to be strictly equal:
// + actual - expected
//
// + 'Hello foobar'
// - 'Hello World!'
//          ^

const apples = 1;
const oranges = 2;
assert.strictEqual(apples, oranges, `apples ${apples} !== oranges ${oranges}`);
// AssertionError [ERR_ASSERTION]: apples 1 !== oranges 2

assert.strictEqual(1, '1', new TypeError('Inputs are not identical'));
// TypeError: Inputs are not identical</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.strictEqual(1, 2);
// AssertionError [ERR_ASSERTION]: Expected inputs to be strictly equal:
//
// 1 !== 2

assert.strictEqual(1, 1);
// OK

assert.strictEqual('Hello foobar', 'Hello World!');
// AssertionError [ERR_ASSERTION]: Expected inputs to be strictly equal:
// + actual - expected
//
// + 'Hello foobar'
// - 'Hello World!'
//          ^

const apples = 1;
const oranges = 2;
assert.strictEqual(apples, oranges, `apples ${apples} !== oranges ${oranges}`);
// AssertionError [ERR_ASSERTION]: apples 1 !== oranges 2

assert.strictEqual(1, '1', new TypeError('Inputs are not identical'));
// TypeError: Inputs are not identical</pre> <p>If the values are not strictly equal, an <a href="#class-assertassertionerror"><code>AssertionError</code></a> is thrown with a <code>message</code> property set equal to the value of the <code>message</code> parameter. If the <code>message</code> parameter is undefined, a default error message is assigned. If the <code>message</code> parameter is an instance of an <a href="errors#class-error"><code>Error</code></a> then it will be thrown instead of the <a href="#class-assertassertionerror"><code>AssertionError</code></a>.</p> </section><section><h3 id="assertthrowsfn-error-message">
<code>assert.throws(fn[, error][, message])</code>
</h3> <div class="api_metadata"> <details class="changelog"><summary>History</summary> <table> <tbody>
<tr>
<th>Version</th>
<th>Changes</th>
</tr> <tr>
<td>v10.2.0</td> <td><p>The <code>error</code> parameter can be an object containing regular expressions now.</p></td>
</tr> <tr>
<td>v9.9.0</td> <td><p>The <code>error</code> parameter can now be an object as well.</p></td>
</tr> <tr>
<td>v4.2.0</td> <td><p>The <code>error</code> parameter can now be an arrow function.</p></td>
</tr> <tr>
<td>v0.1.21</td> <td><p><span>Added in: v0.1.21</span></p></td>
</tr> </tbody>
</table> </details> </div> <ul> <li>
<code>fn</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a>
</li> <li>
<code>error</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp" class="type">&lt;RegExp&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function" class="type">&lt;Function&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object" class="type">&lt;Object&gt;</a> | <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error" class="type">&lt;Error&gt;</a>
</li> <li>
<code>message</code> <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;string&gt;</a>
</li> </ul> <p>Expects the function <code>fn</code> to throw an error.</p> <p>If specified, <code>error</code> can be a <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes"><code>Class</code></a>, <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions"><code>RegExp</code></a>, a validation function, a validation object where each property will be tested for strict deep equality, or an instance of error where each property will be tested for strict deep equality including the non-enumerable <code>message</code> and <code>name</code> properties. When using an object, it is also possible to use a regular expression, when validating against a string property. See below for examples.</p> <p>If specified, <code>message</code> will be appended to the message provided by the <code>AssertionError</code> if the <code>fn</code> call fails to throw or in case the error validation fails.</p> <p>Custom validation object/error instance:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

const err = new TypeError('Wrong value');
err.code = 404;
err.foo = 'bar';
err.info = {
  nested: true,
  baz: 'text'
};
err.reg = /abc/i;

assert.throws(
  () =&gt; {
    throw err;
  },
  {
    name: 'TypeError',
    message: 'Wrong value',
    info: {
      nested: true,
      baz: 'text'
    }
    // Only properties on the validation object will be tested for.
    // Using nested objects requires all properties to be present. Otherwise
    // the validation is going to fail.
  }
);

// Using regular expressions to validate error properties:
assert.throws(
  () =&gt; {
    throw err;
  },
  {
    // The `name` and `message` properties are strings and using regular
    // expressions on those will match against the string. If they fail, an
    // error is thrown.
    name: /^TypeError$/,
    message: /Wrong/,
    foo: 'bar',
    info: {
      nested: true,
      // It is not possible to use regular expressions for nested properties!
      baz: 'text'
    },
    // The `reg` property contains a regular expression and only if the
    // validation object contains an identical regular expression, it is going
    // to pass.
    reg: /abc/i
  }
);

// Fails due to the different `message` and `name` properties:
assert.throws(
  () =&gt; {
    const otherErr = new Error('Not found');
    // Copy all enumerable properties from `err` to `otherErr`.
    for (const [key, value] of Object.entries(err)) {
      otherErr[key] = value;
    }
    throw otherErr;
  },
  // The error's `message` and `name` properties will also be checked when using
  // an error as validation object.
  err
);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

const err = new TypeError('Wrong value');
err.code = 404;
err.foo = 'bar';
err.info = {
  nested: true,
  baz: 'text'
};
err.reg = /abc/i;

assert.throws(
  () =&gt; {
    throw err;
  },
  {
    name: 'TypeError',
    message: 'Wrong value',
    info: {
      nested: true,
      baz: 'text'
    }
    // Only properties on the validation object will be tested for.
    // Using nested objects requires all properties to be present. Otherwise
    // the validation is going to fail.
  }
);

// Using regular expressions to validate error properties:
assert.throws(
  () =&gt; {
    throw err;
  },
  {
    // The `name` and `message` properties are strings and using regular
    // expressions on those will match against the string. If they fail, an
    // error is thrown.
    name: /^TypeError$/,
    message: /Wrong/,
    foo: 'bar',
    info: {
      nested: true,
      // It is not possible to use regular expressions for nested properties!
      baz: 'text'
    },
    // The `reg` property contains a regular expression and only if the
    // validation object contains an identical regular expression, it is going
    // to pass.
    reg: /abc/i
  }
);

// Fails due to the different `message` and `name` properties:
assert.throws(
  () =&gt; {
    const otherErr = new Error('Not found');
    // Copy all enumerable properties from `err` to `otherErr`.
    for (const [key, value] of Object.entries(err)) {
      otherErr[key] = value;
    }
    throw otherErr;
  },
  // The error's `message` and `name` properties will also be checked when using
  // an error as validation object.
  err
);</pre> <p>Validate instanceof using constructor:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.throws(
  () =&gt; {
    throw new Error('Wrong value');
  },
  Error
);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.throws(
  () =&gt; {
    throw new Error('Wrong value');
  },
  Error
);</pre> <p>Validate error message using <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions"><code>RegExp</code></a>:</p> <p>Using a regular expression runs <code>.toString</code> on the error object, and will therefore also include the error name.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.throws(
  () =&gt; {
    throw new Error('Wrong value');
  },
  /^Error: Wrong value$/
);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.throws(
  () =&gt; {
    throw new Error('Wrong value');
  },
  /^Error: Wrong value$/
);</pre> <p>Custom error validation:</p> <p>The function must return <code>true</code> to indicate all internal validations passed. It will otherwise fail with an <a href="#class-assertassertionerror"><code>AssertionError</code></a>.</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

assert.throws(
  () =&gt; {
    throw new Error('Wrong value');
  },
  (err) =&gt; {
    assert(err instanceof Error);
    assert(/value/.test(err));
    // Avoid returning anything from validation functions besides `true`.
    // Otherwise, it's not clear what part of the validation failed. Instead,
    // throw an error about the specific validation that failed (as done in this
    // example) and add as much helpful debugging information to that error as
    // possible.
    return true;
  },
  'unexpected error'
);</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

assert.throws(
  () =&gt; {
    throw new Error('Wrong value');
  },
  (err) =&gt; {
    assert(err instanceof Error);
    assert(/value/.test(err));
    // Avoid returning anything from validation functions besides `true`.
    // Otherwise, it's not clear what part of the validation failed. Instead,
    // throw an error about the specific validation that failed (as done in this
    // example) and add as much helpful debugging information to that error as
    // possible.
    return true;
  },
  'unexpected error'
);</pre> <p><code>error</code> cannot be a string. If a string is provided as the second argument, then <code>error</code> is assumed to be omitted and the string will be used for <code>message</code> instead. This can lead to easy-to-miss mistakes. Using the same message as the thrown error message is going to result in an <code>ERR_AMBIGUOUS_ARGUMENT</code> error. Please read the example below carefully if using a string as the second argument gets considered:</p> <p class="module-type">MJS modules</p>
<pre data-language="js">import assert from 'node:assert/strict';

function throwingFirst() {
  throw new Error('First');
}

function throwingSecond() {
  throw new Error('Second');
}

function notThrowing() {}

// The second argument is a string and the input function threw an Error.
// The first case will not throw as it does not match for the error message
// thrown by the input function!
assert.throws(throwingFirst, 'Second');
// In the next example the message has no benefit over the message from the
// error and since it is not clear if the user intended to actually match
// against the error message, Node.js throws an `ERR_AMBIGUOUS_ARGUMENT` error.
assert.throws(throwingSecond, 'Second');
// TypeError [ERR_AMBIGUOUS_ARGUMENT]

// The string is only used (as message) in case the function does not throw:
assert.throws(notThrowing, 'Second');
// AssertionError [ERR_ASSERTION]: Missing expected exception: Second

// If it was intended to match for the error message do this instead:
// It does not throw because the error messages match.
assert.throws(throwingSecond, /Second$/);

// If the error message does not match, an AssertionError is thrown.
assert.throws(throwingFirst, /Second$/);
// AssertionError [ERR_ASSERTION]</pre>
<p class="module-type">CJS modules</p>
<pre data-language="js">const assert = require('node:assert/strict');

function throwingFirst() {
  throw new Error('First');
}

function throwingSecond() {
  throw new Error('Second');
}

function notThrowing() {}

// The second argument is a string and the input function threw an Error.
// The first case will not throw as it does not match for the error message
// thrown by the input function!
assert.throws(throwingFirst, 'Second');
// In the next example the message has no benefit over the message from the
// error and since it is not clear if the user intended to actually match
// against the error message, Node.js throws an `ERR_AMBIGUOUS_ARGUMENT` error.
assert.throws(throwingSecond, 'Second');
// TypeError [ERR_AMBIGUOUS_ARGUMENT]

// The string is only used (as message) in case the function does not throw:
assert.throws(notThrowing, 'Second');
// AssertionError [ERR_ASSERTION]: Missing expected exception: Second

// If it was intended to match for the error message do this instead:
// It does not throw because the error messages match.
assert.throws(throwingSecond, /Second$/);

// If the error message does not match, an AssertionError is thrown.
assert.throws(throwingFirst, /Second$/);
// AssertionError [ERR_ASSERTION]</pre> <p>Due to the confusing error-prone notation, avoid a string as the second argument.</p></section><div class="_attribution">
  <p class="_attribution-p">
    © Joyent, Inc. and other Node contributors<br>Licensed under the MIT License.<br>Node.js is a trademark of Joyent, Inc. and is used with its permission.<br>We are not endorsed by or affiliated with Joyent.<br>
    <a href="https://nodejs.org/api/assert.html" class="_attribution-link">https://nodejs.org/api/assert.html</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
